048-avoid-broker-restarts-when-in-recovery.md

```diff
@@ -65,24 +69,19 @@ The proposed flow of KafkaRoller
 
 Currently, if a Kafka pod does not become ready within the operational timeout, it checks how many times the pod has been retried to roll. If it hasn't reached the maximum number of attempts (6 by default and set to 10 by KafkaReconcilor), it reschedules the rolling of the pod by appending it to the list. Otherwise, reconciliation fails.
 
-As described in the graph above, the following steps are added before the retry step:
+As described in the graph above, the following steps are added after checking the connection to the broker:
 
-1. Ensure that the Kafka pod is not in a stuck state. The existing `isPodStuck()` can be used for this. If the pod is stuck, continue with the current behaviour to retry rolling of the pod.
-2. Use the Pod DNS name to connect to the KafkaAgent API endpoint. If it cannot connect to the endpoint, continue with the current behaviour to retry rolling of the pod. (The connection might timeout or refused because KafkaAgent is stuck or running older version).
+1. If the AdminClient connection failed, check the broker state instead of force restarting the pod.
+2. Use the Pod DNS name to connect to the KafkaAgent API endpoint. If it cannot connect to the endpoint (the connection might timeout or refused because KafkaAgent is stuck or running older version), continue with the current behaviour.
```

Comment:
Isn't it easier to check if the broker is running but not ready and then check the Agent instead of trying to connect with Admin API first??

Reply:
Yes that could be simpler. We could check the broker state after the initial wait for readiness times out (https://github.com/strimzi/strimzi-kafka-operator/blob/main/cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/KafkaRoller.java#L371). That would mean pod is running but not ready. If broker is in log recovery after the timeout, throw an exception to skip on rest of the checks. Then next time we retry the pod, again would wait for readiness and check the broker state after the timeout. 

Initially I was going to check the broker state here https://github.com/strimzi/strimzi-kafka-operator/blob/main/cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/KafkaRoller.java#L512 but I think what you suggested makes more sense.